aws-codeguruprofiler-profilinggroup/src/test/java/software/amazon/codeguruprofiler/profilinggroup/CreateHandlerTest.java

```diff
@@ -43,83 +53,226 @@
-    }
+        @Test
+        public void testSuccess() {
+            final ProgressEvent<ResourceModel, CallbackContext> response = subject.handleRequest(proxy, request, null, logger);
 
-    @Test
-    public void testInternalServerException() {
-        doThrow(InternalServerException.builder().build())
-                .when(proxy).injectCredentialsAndInvokeV2(any(), any());
+            assertResponse(response);
+        }
 
-        assertThrows(CfnServiceInternalErrorException.class, () ->
-                new CreateHandler().handleRequest(proxy, request, null, logger));
+        @Test
+        public void testCorrectOperationIsCalled() {
+            subject.handleRequest(proxy, request, null, logger);
+
+            verify(proxy, times(1)).injectCredentialsAndInvokeV2(eq(createPgRequest), any());
+            verifyNoMoreInteractions(proxy);
+        }
     }
 
-    @Test
-    public void testServiceQuotaExceededException() {
-        doThrow(ServiceQuotaExceededException.builder().build())
-                .when(proxy).injectCredentialsAndInvokeV2(any(), any());
+    @Nested
+    class WithPermissions {
+
+        @BeforeEach
+        public void setup() {
+            doReturn(CreateProfilingGroupResponse.builder().build())
+                    .when(proxy).injectCredentialsAndInvokeV2(any(CreateProfilingGroupRequest.class), any());
+        }
+
+        @Nested
+        class NoPrincipals {
+
+            @Test
+            public void testNullPermissions() {
+                ResourceModel model = newResourceModel(null);
+                request = makeRequest(model);
+                assertResponse(subject.handleRequest(proxy, request, null, logger));
+            }
+
+            @Test
+            public void testNullPrincipals() {
+                ResourceModel model = newResourceModel(AgentPermissions.builder().principals(null).build());
+                request = makeRequest(model);
+                assertResponse(subject.handleRequest(proxy, request, null, logger));
+            }
+
+            @AfterEach
+            public void testCorrectOperationIsCalled() {
+                verify(proxy, times(1)).injectCredentialsAndInvokeV2(eq(createPgRequest), any());
+                verifyNoMoreInteractions(proxy);
+            }
+        }
+
+        @Nested
+        class WithPrincipals {
+
+            @BeforeEach
+            public void setup() {
+                ResourceModel model = newResourceModel(AgentPermissions.builder().principals(Arrays.asList("a", "bc")).build());
+                request = makeRequest(model);
+            }
+
+            @Test
+            public void testSuccess() {
+                assertResponse(subject.handleRequest(proxy, request, null, logger));
+                verify(proxy, times(1)).injectCredentialsAndInvokeV2(eq(createPgRequest), any());
+                verify(proxy, times(1)).injectCredentialsAndInvokeV2(any(PutPermissionRequest.class), any());
```

Comment:
Can we validate the specific request here? This check will still pass if we create the request wrongly.